homeassistant/components/switch/mqtt.py

```diff
@@ -63,81 +62,93 @@
     async def async_discover(discovery_payload):
         """Discover and add a MQTT switch."""
         config = PLATFORM_SCHEMA(discovery_payload)
-        await _async_setup_entity(hass, config, async_add_entities,
+        await _async_setup_entity(config, async_add_entities,
                                   discovery_payload[ATTR_DISCOVERY_HASH])
 
     async_dispatcher_connect(
         hass, MQTT_DISCOVERY_NEW.format(switch.DOMAIN, 'mqtt'),
         async_discover)
 
 
-async def _async_setup_entity(hass, config, async_add_entities,
+async def _async_setup_entity(config, async_add_entities,
                               discovery_hash=None):
     """Set up the MQTT switch."""
-    value_template = config.get(CONF_VALUE_TEMPLATE)
-    if value_template is not None:
-        value_template.hass = hass
-
-    newswitch = MqttSwitch(
-        config.get(CONF_NAME),
-        config.get(CONF_ICON),
-        config.get(CONF_STATE_TOPIC),
-        config.get(CONF_COMMAND_TOPIC),
-        config.get(CONF_AVAILABILITY_TOPIC),
-        config.get(CONF_QOS),
-        config.get(CONF_RETAIN),
-        config.get(CONF_PAYLOAD_ON),
-        config.get(CONF_PAYLOAD_OFF),
-        config.get(CONF_STATE_ON),
-        config.get(CONF_STATE_OFF),
-        config.get(CONF_OPTIMISTIC),
-        config.get(CONF_PAYLOAD_AVAILABLE),
-        config.get(CONF_PAYLOAD_NOT_AVAILABLE),
-        config.get(CONF_UNIQUE_ID),
-        value_template,
-        config.get(CONF_DEVICE),
-        discovery_hash,
-    )
-
-    async_add_entities([newswitch])
+    async_add_entities([MqttSwitch(config, discovery_hash)])
 
 
 # pylint: disable=too-many-ancestors
 class MqttSwitch(MqttAvailability, MqttDiscoveryUpdate, MqttEntityDeviceInfo,
                  SwitchDevice, RestoreEntity):
     """Representation of a switch that can be toggled using MQTT."""
 
-    def __init__(self, name, icon,
-                 state_topic, command_topic, availability_topic,
-                 qos, retain, payload_on, payload_off, state_on,
-                 state_off, optimistic, payload_available,
-                 payload_not_available, unique_id: Optional[str],
-                 value_template, device_config: Optional[ConfigType],
-                 discovery_hash):
+    def __init__(self, config, discovery_hash):
         """Initialize the MQTT switch."""
-        MqttAvailability.__init__(self, availability_topic, qos,
+        self._state = False
+        self._sub_state = None
+
+        self._name = None
```

Comment:
This should be updated to just store `config` on the instance

Reply:
@balloob Sure, I'll open a follow-up which does that once this is merged.
Let me know if you prefer to include those changes here as well.